torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -2760,64 +2770,73 @@ std::shared_ptr<NCCLComm> ProcessGroupNCCL::initNCCLComm(
-  auto ranksPerRoot = getCvarInt(TORCH_NCCL_SCALABLE_INIT_RANKS_PER_ROOT, 128);
-  auto numRoots = (getSize() / ranksPerRoot) + (getSize() % ranksPerRoot > 0);
-  std::vector<ncclUniqueId> ncclIDs(numRoots);
-
-  if (!ncclComm) {
-    C10D_NCCL_CHECK(ncclGetUniqueId(&ncclID), std::nullopt);
-    // We only need to all-gather the ncclID if the rank is root.
-    auto timeStarted = std::chrono::steady_clock::now();
-    allgatherUniqueNCCLID(
-        getRootRank(rank_, getSize(), numRoots), &ncclID, ncclIDs);
-    auto timerDeltaMs =
-        std::chrono::duration_cast<std::chrono::duration<double>>(
-            std::chrono::steady_clock::now() - timeStarted)
-            .count() *
-        1000;
-    LOG(INFO) << logPrefix()
-              << "ProcessGroupNCCL broadcast unique ID through store took "
-              << timerDeltaMs << " ms";
-    ncclComm =
-        NCCLComm::create_scalable(numRanks, rank, ncclIDs, options_->config);
-  }
-#else
-  // To simplify conditional nesting, just create the ncclComms[i]
-  // entry if it hasn't been yet rather than untangling the
-  // conditions that might have resulted in a split above.
-  if (!ncclComm) {
-    if (getCvarBool(TORCH_NCCL_BCAST_UNIQUEID, true) && !isSendRecvSelf) {
-      // For point-to-point communication, lower rank of the two will get unique
-      // id.
-      if (rank_ == 0 || (singleP2POp && p2pRank == 0)) {
+  auto ranksPerRoot = getCvarInt(TORCH_NCCL_RANKS_PER_ROOT, 128);
+#ifdef NCCL_HAS_INIT_RANK_SCALABLE
+  useScalableInit = !singleP2POp && (getSize() > ranksPerRoot);
+#endif // NCCL_HAS_INIT_RANK_SCALABLE
+
+  if (useScalableInit) {
+    auto numRoots = (getSize() + ranksPerRoot - 1) / ranksPerRoot;
+    std::vector<ncclUniqueId> ncclIDs(numRoots);
+
+    if (!ncclComm) {
+      auto rootRankAndIndex = getRootRankAndIndex(rank_, getSize(), numRoots);
+      // We only need to get unique IDs for roots.
+      if (rank_ == rootRankAndIndex.first) {
         C10D_NCCL_CHECK(ncclGetUniqueId(&ncclID), std::nullopt);
       }
-
-      // Broadcast so that each process can have a unique NCCL ID
+      // We only need to all-gather the ncclID if the rank is root.
       auto timeStarted = std::chrono::steady_clock::now();
-      broadcastUniqueNCCLID(&ncclID, singleP2POp, deviceKey, p2pRank);
+      allgatherUniqueNCCLIDs(
+          rootRankAndIndex.first, rootRankAndIndex.second, &ncclID, ncclIDs);
       auto timerDeltaMs =
           std::chrono::duration_cast<std::chrono::duration<double>>(
               std::chrono::steady_clock::now() - timeStarted)
               .count() *
           1000;
       LOG(INFO) << logPrefix()
-                << "ProcessGroupNCCL broadcast unique ID through store took "
+                << "ProcessGroupNCCL all-gather unique IDs through store took "
                 << timerDeltaMs << " ms";
+      ncclComm = NCCLComm::create_scalable(
+          numRanks, rootRankAndIndex.second, rank, ncclIDs, options_->config);
```

Comment:
nit: you may have to either:
- add `#ifdef` for `create_scalable`, with a `#else` throwing an internal error. 
OR
- declare `create_scalable` unconditionally in NCCLUtils.hpp.

Reply:
oh... no wonder It has errors.